File: scripts/textures/horizon-map.test.ts
import { readdirSync, readFileSync } from 'node:fs';
import { resolve } from 'node:path';

import { describe, expect, it } from 'vitest';

import { SOL_BODIES } from '../../src/client/solar-system/planet-system';
import {
  HORIZON_AZIMUTHS,
  HORIZON_SIN_RANGE,
  RELIEF_ELEV_SPAN_M,
  reliefHorizonSines,
} from '../../src/client/solar-system/planets/surface-relief/surface-relief-pure';
import {
  LIMB_EXP,
  LIMB_FLOOR,
} from '../../src/client/solar-system/planets/emission/mesh-surface-pure';
import { isLfsPointer, webpSize } from './image-header-pure';

// horizon_map.py cannot import the runtime tables, so this pins its constants
// back against the shader's mirror of them, and both against the shipped
// artifacts. Why it matters: data/textures/README.md § Cast shadows.

const TEXTURES = resolve(__dirname, '../../data/textures');
const MESH_FRAG = resolve(
  __dirname,
  '../../src/client/solar-system/planets/planet-mesh.frag.glsl',
);
const pySource = readFileSync(resolve(__dirname, 'horizon_map.py'), 'utf-8');
const buildSource = readFileSync(resolve(__dirname, 'build-textures.py'), 'utf-8');

interface HorizonRow {
  width: number;
  azimuths: number;
  medianHorizonDeg: number;
  p99HorizonDeg: number;
  clampedPct: number;
}

const manifest: Record<string, { horizon?: HorizonRow }> = JSON.parse(
  readFileSync(resolve(TEXTURES, 'relief.json'), 'utf-8'),
);

const pyNumber = (name: string): number => {
  const m = pySource.match(new RegExp(`^${name} = ([\\d.]+)$`, 'm'));
  expect(m, name).not.toBeNull();
  return Number(m![1]);
};

const HALVES = ['a', 'b'] as const;
const planePath = (body: string, half: string) =>
  resolve(TEXTURES, `${body}-horizon-${half}.webp`);

const halvesOnDisk = readdirSync(TEXTURES).filter((f) =>
  /-horizon-[ab]\.webp$/.test(f),
);
const bodiesWithAnyHalf = [
  ...new Set(halvesOnDisk.map((f) => f.replace(/-horizon-[ab]\.webp$/, ''))),
].sort();
/** Only bodies with BOTH halves, so nothing below reads a file that is not
 *  there — a body holding one half is the assertion's job, not an ENOENT. */
const shipped = bodiesWithAnyHalf.filter((body) =>
  HALVES.every((half) => halvesOnDisk.includes(`${body}-horizon-${half}.webp`)),
);

const mapsArePointers = shipped.some((body) =>
  HALVES.some((half) => isLfsPointer(readFileSync(planePath(body, half)))),
);
if (mapsArePointers) {
  console.warn(
    '[horizon-map] skipping artifact-header pins — horizon maps are LFS ' +
      'pointers, not WebP. Run `git lfs pull` to exercise them.',
  );
}

describe('horizon maps', () => {
  it('ships both halves for exactly the relief bodies', () => {
    // Every azimuth or none: the shader interpolates across the seam between
    // the two planes, so a body with one is a body with a wrong skyline.
    expect(bodiesWithAnyHalf).toEqual(shipped);
    expect(shipped).toEqual(Object.keys(RELIEF_ELEV_SPAN_M).sort());
  });

  it('splits the azimuths evenly over the two planes', () => {
    expect(pyNumber('HORIZON_AZIMUTHS')).toBe(HORIZON_AZIMUTHS);
    expect(HORIZON_AZIMUTHS % 4).toBe(0);
    expect(HORIZON_AZIMUTHS / 4).toBe(HALVES.length);
  });

  it('encodes the sine on the scale the shader decodes with', () => {
    expect(pyNumber('HORIZON_SIN_RANGE')).toBe(HORIZON_SIN_RANGE);
  });

  it('reaches past every body limb bound, which is where the floor sits', () => {
    // The negative half of the encoding only has to hold the deepest a skyline
    // can fall, and that is the body's own limb; the positive half is what
    // needs the headroom, for a crater wall seen from the floor.
    for (const name of Object.keys(RELIEF_ELEV_SPAN_M)) {
      const body = SOL_BODIES.find((b) => b.name.toLowerCase() === name)!;
      const [, none] = reliefHorizonSines(RELIEF_ELEV_SPAN_M[name], body.radiusKm);
      expect(none, `${name} limb bound`).toBeLessThan(HORIZON_SIN_RANGE);
    }
  });

  it('searches exactly as far as that same limb bound', () => {
    // search_arc IS arccos(r_floor / r_summit), whose sine is the `none at all`
    // bound the shader falls back to — one quantity, so a body that changes its
    // DEM span cannot leave the precompute searching the old distance.
    expect(pySource).toContain(
      'np.arccos((r + spec["span_m"][0]) / (r + spec["span_m"][1]))');
    for (const name of Object.keys(RELIEF_ELEV_SPAN_M)) {
      const body = SOL_BODIES.find((b) => b.name.toLowerCase() === name)!;
      const [, none] = reliefHorizonSines(RELIEF_ELEV_SPAN_M[name], body.radiusKm);
      const r = body.radiusKm * 1000;
      const span = RELIEF_ELEV_SPAN_M[name];
      const arc = Math.acos((r + span[0]) / (r + span[1]));
      expect(Math.sin(arc), `${name}`).toBeCloseTo(none, 12);
    }
  });

  it('starts the march two output texels out and steps at the DEM resolution', () => {
    // Two independent choices. The start distance keeps the map off ground the
    // colour and normal maps own — inside one output texel a caster is half a
    // colour-map texel, so nothing on screen can throw the shadow. The step
    // stays at the DEM's resolution so a narrow ridge at range is sampled
    // rather than averaged away (README.md § Cast shadows).
    expect(pyNumber('HORIZON_MARCH_START_TEXELS')).toBe(2);
    expect(pySource).toContain(
      'return HORIZON_MARCH_START_TEXELS * 2 * np.pi / out_width');
    expect(pySource).toContain('step = 2 * np.pi / w_d');
    expect(pySource).toContain(
      'steps = max(2, int(np.ceil((psi_max - psi_min) / step)) + 1)');
  });

  it('starts past the caster the colour map could resolve, on every body', () => {
    // The start distance is set in output texels, so the kilometres it buys
    // differ per body — this is the number that has to clear a colour-map texel
    // for the shadow to have a visible caster.
    const startKm: Record<string, number> = { moon: 10.7, mercury: 15.0, mars: 20.8 };
    const texels = pyNumber('HORIZON_MARCH_START_TEXELS');
    const width = pyNumber('HORIZON_TARGET_W');
    expect(Object.keys(startKm).sort()).toEqual(shipped);
    // Below two the nearest caster sits inside a single colour-map texel again,
    // which is the whole defect this distance exists to fix.
    expect(texels).toBeGreaterThanOrEqual(2);
    for (const [name, km] of Object.entries(startKm)) {
      const body = SOL_BODIES.find((b) => b.name.toLowerCase() === name)!;
      expect((texels * 2 * Math.PI * body.radiusKm) / width, name).toBeCloseTo(km, 1);
    }
  });

  it('registers each map to the colour map, like its normal map', () => {
    expect(pySource).toContain('elev = roll_to_map_centre(elev, spec)');
  });

  it('merges both halves of a body relief row instead of replacing it', () => {
    // The normal map and the horizon pair sit behind separate up-to-date gates
    // with different dependency sets, so either can be skipped while the other
    // runs. An assignment on either side drops the skipped half's stats from
    // relief.json, and the manifest pins below are what would then fail.
    expect(buildSource).toContain('relief.setdefault(name, {}).update(stats)');
    expect(buildSource).toContain('relief.setdefault(name, {})["horizon"] = stats');
    expect(buildSource).not.toMatch(/^\s*rgb, relief\[name\] = /m);
  });

  it('writes the planes with libwebp exact, so alpha cannot eat RGB', () => {
    // Without it libwebp is free to rewrite RGB wherever alpha is 0, which here
    // is one azimuth's skyline silently overwriting three others.
    expect(buildSource).toContain('out_path, "WEBP", lossless=True, exact=True');
  });

  it('records the achieved width, azimuths and clamping for every body', () => {
    const pins: Record<string, [number, number, number]> = {
      moon: [0.556, 10.835, 0.0002],
      mercury: [0.057, 4.861, 0.0],
      mars: [-0.141, 3.428, 0.0],
    };
    const width = pyNumber('HORIZON_TARGET_W');
    expect(Object.keys(pins).sort()).toEqual(shipped);
    for (const [name, [median, p99, clamped]] of Object.entries(pins)) {
      const row = manifest[name].horizon!;
      expect(row.width, `${name} width`).toBe(width);
      expect(row.azimuths, `${name} azimuths`).toBe(HORIZON_AZIMUTHS);
      expect(row, `${name} stats`).toMatchObject({
        medianHorizonDeg: median,
        p99HorizonDeg: p99,
        clampedPct: clamped,
      });
    }
  });

  it('keeps the Moon the dominant skyline, as it is the dominant relief', () => {
    const p99 = (name: string) => manifest[name].horizon!.p99HorizonDeg;
    expect(p99('moon')).toBeGreaterThan(2 * p99('mercury'));
    expect(p99('mercury')).toBeGreaterThan(p99('mars'));
  });

  it('measures the disc against the shading the renderer actually does', () => {
    // measure_relief_lighting.py integrates the disc to decide whether relief
    // needs a flux renormalisation (../../src/client/solar-system/planets/
    // emission/README.md). Its own copies of the limb constants drifting from
    // the shader's would answer that question about a different body.
    const measure = readFileSync(resolve(__dirname, 'measure_relief_lighting.py'), 'utf-8');
    expect(measure).toContain(`LIMB_FLOOR = ${LIMB_FLOOR}`);
    expect(measure).toContain(`LIMB_EXP = ${LIMB_EXP}`);
    // The third mirrored constant is a shader literal rather than a TS export,
    // so it pins against the GLSL directly instead of an import.
    const floor = measure.match(/^TERM_SOFTNESS_FLOOR = (\S+)$/m);
    expect(floor, 'TERM_SOFTNESS_FLOOR').not.toBeNull();
    expect(readFileSync(MESH_FRAG, 'utf-8')).toContain(
      `max(uTermSoftness, ${floor![1]})`);
  });

  it('mirrors each body albedo the interreflected term pays twice', () => {
    // The fourth column of the phase table is the only place the fill term's
    // flux is measured, and it scales linearly with this number — a stale copy
    // would answer the renormalisation question about a different body.
    const measure = readFileSync(resolve(__dirname, 'measure_relief_lighting.py'), 'utf-8');
    const row = measure.match(/^ALBEDO = \{(.+)\}$/m);
    expect(row, 'ALBEDO').not.toBeNull();
    const mirrored = Object.fromEntries(
      [...row![1].matchAll(/"(\w+)": ([\d.]+)/g)].map(([, k, v]) => [k, Number(v)]),
    );
    expect(Object.keys(mirrored).sort()).toEqual(shipped);
    for (const [name, albedo] of Object.entries(mirrored)) {
      const body = SOL_BODIES.find((b) => b.name.toLowerCase() === name)!;
      expect(albedo, `${name} albedo`).toBe(body.albedo);
    }
  });

  it.skipIf(mapsArePointers)('ships both planes at the declared width', () => {
    const width = pyNumber('HORIZON_TARGET_W');
    for (const body of shipped) {
      for (const half of HALVES) {
        expect(
          webpSize(readFileSync(planePath(body, half)), `${body}-${half}`),
          `${body}-horizon-${half}`,
        ).toEqual({ width, height: width / 2 });
      }
    }
  });
});

File: scripts/textures/horizon_map.test.py
#!/usr/bin/env python3
"""Stdlib-unittest pins for horizon_map.py: the spherical elevation-angle
geometry, azimuth registration, the limb identity and the encoding round-trip.
Run directly: python3 horizon_map.test.py (needs NumPy)."""

from __future__ import annotations

import math
import sys
import unittest
from pathlib import Path

import numpy as np

sys.path.insert(0, str(Path(__file__).resolve().parent))
from dem_relief import DEM_BODIES  # noqa: E402
from horizon_map import (  # noqa: E402
    HORIZON_AZIMUTHS,
    HORIZON_MARCH_START_TEXELS,
    HORIZON_SIN_RANGE,
    HORIZON_TARGET_W,
    decode_horizon_sin,
    encode_horizon,
    horizon_angles,
    march_start,
    search_arc,
)

# Narrow enough to run in a second. The output width is the DEM width in every
# case below, so registration is read exactly rather than through box_reduce.
W = 512
H = W // 2
MOON = DEM_BODIES["moon"]
R_MOON = MOON["radius_km"] * 1000.0

# Azimuth 0 is east and they run toward north — the order the two shipped
# planes concatenate in, and the order the mesh shader indexes.
AZIMUTHS = ("E", "NE", "N", "NW", "W", "SW", "S", "SE")


def limb_sin(ground_m: float, summit_m: float) -> float:
    """`limbSin` in surface-relief-pure.ts: sine of the depression ground at
    `summit_m` can still see the sun at, over terrain at `ground_m`."""
    return math.sqrt(max(0.0, 1 - (ground_m / summit_m) ** 2))


def flat_floor(out_width: int = W) -> float:
    """Horizon angle of flat ground at the reference sphere: the sphere falling
    away over the march's start distance, and NOT zero. It is a maximum over a
    ray whose angle decreases monotonically outward, so the start distance is
    the whole answer and nothing further along the march can raise it."""
    psi = march_start(out_width)
    return math.atan2(math.cos(psi) - 1, math.sin(psi))


class SearchArcTests(unittest.TestCase):
    def test_is_the_renderers_fallback_limb_bound(self) -> None:
        """One quantity: a body that changes its DEM span cannot leave the
        precompute searching the distance the old span implied."""
        for name, spec in DEM_BODIES.items():
            r = spec["radius_km"] * 1000.0
            expected = limb_sin(r + spec["span_m"][0], r + spec["span_m"][1])
            self.assertAlmostEqual(
                math.sin(search_arc(spec)), expected, places=12, msg=name
            )

    def test_covers_hundreds_of_kilometres(self) -> None:
        arcs_km = {
            name: search_arc(spec) * spec["radius_km"]
            for name, spec in DEM_BODIES.items()
        }
        self.assertEqual(
            {k: round(v) for k, v in arcs_km.items()},
            {"moon": 262, "mercury": 219, "mars": 446},
        )


class FlatGroundTests(unittest.TestCase):
    def test_reads_the_curvature_drop_over_the_start_distance(self) -> None:
        """Flat ground reads the march's start distance's own drop, not 0 — the
        sun stays up that far past the geometric terminator. Bounded, and it is
        why the composition's coarse-map slack runs toward lighting, never
        shadowing."""
        angles = horizon_angles(np.zeros((H, W), np.float32), MOON, 8, W)
        expected = flat_floor()
        self.assertAlmostEqual(math.degrees(expected), -0.703125, places=5)
        self.assertEqual(np.ptp(angles), 0.0)
        self.assertAlmostEqual(float(angles[0, 0, 0]), expected, places=6)

    def test_the_floor_follows_the_output_grid_not_the_dem_step(self) -> None:
        """The two march parameters are independent, and this is the half the
        start distance owns: refining the DEM refines where a narrow ridge is
        sampled along the ray but moves the near bound not at all, so the floor
        is identical. Halving the OUTPUT width doubles it."""
        coarse = horizon_angles(np.zeros((H, W), np.float32), MOON, 8, W)
        fine = horizon_angles(np.zeros((2 * H, 2 * W), np.float32), MOON, 8, W)
        self.assertAlmostEqual(float(fine[0, 0, 0]), float(coarse[0, 0, 0]), places=6)
        wider = horizon_angles(np.zeros((H, W), np.float32), MOON, 8, W // 2)
        self.assertAlmostEqual(float(wider[0, 0, 0]), flat_floor(W // 2), places=6)
        self.assertAlmostEqual(
            float(wider[0, 0, 0]) / float(coarse[0, 0, 0]), 2.0, places=3
        )

    def test_refuses_an_output_grid_coarser_than_the_search_arc(self) -> None:
        """A near bound past the search arc would put every sample inside the
        near field the march exists to skip, silently restoring the defect. The
        Moon's arc is 0.151 rad, so a 32-wide output asks for 0.393."""
        with self.assertRaises(AssertionError):
            horizon_angles(np.zeros((H, W), np.float32), MOON, 8, 32)

    def test_starts_past_what_the_colour_map_can_resolve(self) -> None:
        """The reason for the start distance rather than its consequence: a
        caster inside one output texel is half a colour-map texel, so the ground
        throwing the shadow cannot be drawn at any camera distance."""
        self.assertGreaterEqual(HORIZON_MARCH_START_TEXELS, 2.0)
        km = march_start(HORIZON_TARGET_W) * MOON["radius_km"]
        self.assertAlmostEqual(km, 10.66, places=2)


class AzimuthRegistrationTests(unittest.TestCase):
    """A skyline pinned to the wrong azimuth shades real terrain lit from the
    wrong side and looks entirely plausible doing it, so the handedness is
    pinned here rather than left to the browser."""

    # The output width is the DEM width here, so the march's samples land on
    # whole texels and this puts the wall on the very first one — any closer and
    # it sits inside the near field the march deliberately skips.
    WALL = int(HORIZON_MARCH_START_TEXELS)

    @classmethod
    def setUpClass(cls) -> None:
        cls.row, cls.col = H // 2, W // 2
        elev = np.zeros((H, W), np.float32)
        # Three rows rather than one texel: the diagonal rays leave the row they
        # started on, so a point wall is off their bearing entirely and every
        # azimuth but due east reads the floor. A rim has meridional extent.
        elev[cls.row - 1 : cls.row + 2, cls.col + cls.WALL] = 20000.0
        cls.angles = horizon_angles(elev, MOON, 8, W)
        cls.floor = flat_floor()

    def by_name(self, col: int) -> dict[str, float]:
        return dict(zip(AZIMUTHS, (float(v) for v in self.angles[self.row, col])))

    def test_a_wall_due_east_lands_on_azimuth_zero(self) -> None:
        seen = self.by_name(self.col)
        self.assertEqual(max(seen, key=lambda k: seen[k]), "E")
        self.assertGreater(seen["E"], math.radians(20))

    def test_the_azimuths_it_cannot_be_seen_from_stay_at_the_floor(self) -> None:
        seen = self.by_name(self.col)
        for name in ("N", "NW", "W", "SW", "S"):
            self.assertAlmostEqual(seen[name], self.floor, places=9, msg=name)

    def test_the_two_diagonals_bracketing_east_agree(self) -> None:
        seen = self.by_name(self.col)
        # No output row falls exactly on the equator (they are cell centres of
        # an even count), so this one sits 0.35° south of it and the two
        # diagonals are near-mirrors rather than exact ones.
        self.assertAlmostEqual(seen["NE"], seen["SE"], delta=2e-4)
        self.assertLess(seen["NE"], seen["E"])
        self.assertGreater(seen["NE"], self.floor)

    def test_the_far_side_sees_the_same_wall_due_west(self) -> None:
        """East and west are one axis read from both ends: the texel as far past
        the wall must see it at azimuth 4 exactly as strongly."""
        east_of_wall = self.by_name(self.col + 2 * self.WALL)
        self.assertEqual(
            max(east_of_wall, key=lambda k: east_of_wall[k]), "W"
        )
        self.assertAlmostEqual(
            east_of_wall["W"], self.by_name(self.col)["E"], places=6
        )


class LimbBoundTests(unittest.TestCase):
    def test_a_summit_sees_the_sun_down_to_its_own_limb_bound(self) -> None:
        """The whole point of the height term: an isolated summit reads a
        NEGATIVE skyline, and it is the same bound the shader falls back to
        while the maps load (`reliefHorizonSines`'s `full`)."""
        summit_m = MOON["span_m"][1]
        elev = np.zeros((H, W), np.float32)
        elev[H // 2, W // 2] = summit_m
        angles = horizon_angles(elev, MOON, 8, W)
        bound = math.asin(limb_sin(R_MOON, R_MOON + summit_m))
        self.assertAlmostEqual(math.degrees(bound), 6.360, places=3)
        for a, name in enumerate(AZIMUTHS):
            got = float(angles[H // 2, W // 2, a])
            self.assertLessEqual(got, -bound, msg=name)
            self.assertGreater(got, -bound - math.radians(0.02), msg=name)


class EncodingTests(unittest.TestCase):
    @staticmethod
    def planes_of(angles: np.ndarray) -> np.ndarray:
        first, second = encode_horizon(angles)
        return np.concatenate([first, second], axis=2).astype(np.float32)

    def ramp(self) -> np.ndarray:
        """One distinct sine per azimuth, spread across the encoding range."""
        angles = np.zeros((2, 2, HORIZON_AZIMUTHS), np.float32)
        for k in range(HORIZON_AZIMUTHS):
            angles[..., k] = math.asin((k - 3.5) / 10.0)
        return angles

    def test_splits_the_azimuths_four_and_four(self) -> None:
        first, second = encode_horizon(self.ramp())
        self.assertEqual(first.shape, (2, 2, 4))
        self.assertEqual(second.shape, (2, 2, 4))
        self.assertEqual(first.dtype, np.uint8)

    def test_each_azimuth_decodes_off_its_own_bearing(self) -> None:
        """Round-trips within half a quantisation step, and the bearing → slot
        walk lands on the channel the wall tests put the signal in."""
        planes = self.planes_of(self.ramp())
        lsb = 2 * HORIZON_SIN_RANGE / 255
        bearings = ((1, 0), (1, 1), (0, 1), (-1, 1), (-1, 0), (-1, -1), (0, -1), (1, -1))
        for k, (east, north) in enumerate(bearings):
            got = decode_horizon_sin(
                planes, np.full((2, 2), float(east)), np.full((2, 2), float(north))
            )
            self.assertAlmostEqual(
                float(got[0, 0]), (k - 3.5) / 10.0, delta=lsb / 2, msg=AZIMUTHS[k]
            )

    def test_clamps_beyond_the_encoding_range(self) -> None:
        steep = np.full((1, 1, HORIZON_AZIMUTHS), math.radians(80), np.float32)
        planes = self.planes_of(steep)
        got = decode_horizon_sin(planes, np.ones((1, 1)), np.zeros((1, 1)))
        self.assertAlmostEqual(float(got[0, 0]), HORIZON_SIN_RANGE, places=6)


if __name__ == "__main__":
    unittest.main()
